docs/spec: Proposal for new manifest format

aaronlehmann opened the pull request:
This is a follow-on to PR #62, and it borrows much of the format
from #993, but uses specific formats for the image manifest and manifest
list (fat manifest) instead of a combined generic format.

The intent of this proposed manifest format is to allow multi-arch, and
allow for full content-addressability of images in the Docker engine.

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

aaronlehmann commented:
cc @stevvooe @estesp @tonistiigi @mattmoor

mattmoor commented on docs/spec/manifest-v2-2.md:
```diff
@@ -0,0 +1,213 @@
+- **`layers`** *array*
+
+    The layer list is ordered starting from the base image (opposite order of schema1).
+    The `baseLayer` label is intended to improve human comprehensibility of the layer
```
Perhaps it would be better if the specification left room for this kind of unstructured annotation without calling them out?  This way, different implementations can add/remove information intended for humans / UI without a schema change?

For examples, in the future we might want to know the "top" layer, or per-layer what it's Dockerfile directive was, or...

Maybe this should be "comment": "base layer"?

aaronlehmann replied:
I'm fine with that, or we can avoid calling it out in the spec. The intention was to clarify that the label in the example doesn't have any semantic meaning to the engine.

mattmoor commented on docs/spec/manifest-v2-2.md:
````diff
@@ -0,0 +1,213 @@
+*Example showing a simple manifest list pointing to image manifests for two platforms:*
+```json
+{
+  "schemaVersion": 3,
````
Why 3?

dmcgowan replied:
The new manifest format is already 2, this is really a 3rd format specifically for listing multiple manifests. The schema version must be set such that old clients can fail early and predictability when this manifest list is not supported. Setting this to 2 as well would have been incorrect and confusing.

mattmoor commented on docs/spec/manifest-v2-2.md:
```diff
@@ -0,0 +1,213 @@
+{
+    "schemaVersion": 2,
+    "config": {
+        "mediaType": "application/vnd.docker.container.image.v1+json",
```
At this point, isn't the media type here implied?  What else could it be?

RichardScothern replied:
If the format of an image container evolves, it would have a new schema and corresponding mediatype.

waghpooja replied:
The media type and schema version here go hand in hand. So having both opens up the possibility of these going out of sync as is the case in this example where schemaVersion is 2 but mediaType is application/vnd.docker.container.image.v1+json instead of pplication/vnd.docker.container.image.v2+json.

stevvooe replied:
I think we should phase out `schemaVersion` and rely on the mediatype. The `schemaVersion` field forces a two step deserialization process. We should just have `schemaVersion: 2` as an open keyspace. If it contains a "manifests" key, it has manifests, if it has a "config" key, it has a config. Mediatypes are ultimately more flexible and there is no reason to really carry on with `schemaVersion`, unless we modify the keyspace after 2.

mattmoor commented on docs/spec/manifest-v2-2.md:
```diff
@@ -0,0 +1,213 @@
+    },
+    "layers": [
+        {
+            "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
```
Same question as above.

mattmoor commented on docs/spec/manifest-v2-2.md:
```diff
@@ -0,0 +1,213 @@
+  "schemaVersion": 3,
+  "manifests": [
+    {
+      "mediaType": "application/vnd.docker.image.manifest.v2+json",
```
Can this be any other media type?

aaronlehmann replied:
I was thinking we should support `application/vnd.docker.image.manifest.v2+json` as well, in case someone really wants to reference a legacy manifest here, but I don't have any strong feelings about allowing this, and can see some reasons why we might not want to allow it.

mattmoor commented on docs/spec/manifest-v2-2.md:
```diff
@@ -0,0 +1,213 @@
+- `application/vnd.docker.image.rootfs.diff.tar.gzip`: "Layer", as a gzipped tar
+- `application/vnd.docker.container.image.v1+json`: Container config JSON
+
+## Manifest List
```
This kind of confused me a little bit because today we have endpoints:
```
/v2/<name>/manifests/<reference>
/v2/<name>/tags/list
```

I initially read this as:
```
/v2/<name>/manifests/list
```

Is it worth splitting the resource noun for this from `manifest` to something like `group`?
```
/v2/<name>/groups/<reference>
```
(`label` seemed like the wrong noun because it is used in a different sense within the JSON)

RichardScothern commented on docs/spec/manifest-v2-2.md:
```diff
@@ -0,0 +1,213 @@
+
+    - **`labels`** *object*
+
+        Labels may be keyed to *any* JSON object, allowing content creators to
```
We should forbid them being keyed to themselves.

dmcgowan commented:
We should add a compatibility section at the end which will suggest how we will support clients.

The original suggestion is that clients which support this would provide an "Accept" header and the registry will now set the "Content-Type". When either of these headers are missing the assumption made by the other endpoint would be that schema version 1 is used.

waghpooja commented on docs/spec/manifest-v2-2.md:
```diff
@@ -0,0 +1,213 @@
+
+- **`layers`** *array*
+
+    The layer list is ordered starting from the base image (opposite order of schema1).
```
What is the motivation behind reversing this order from the one in schema1?

aaronlehmann replied:
The ordering in schema1 was awkward because engines ended up having to process it in reverse. The base layer needs to be added to the graph first, but this was at the end of the list in schema1.

estesp commented:
I like this as a cleaner alternate to #993. I assume we still have some questions to answer based on the questions in-line above from @mattmoor and others, but given we can find answers to those, I'm LGTM on this spec design for supporting multi-arch.

rbarlow commented:
Will this format allow there to be a canonical format for manifests so that there is exactly one way to calculate the digest of a manifest? @TomasTomecek has been working on writing a Python library to calculate the digest of a manifest which has been tricky due to manifests not having a "correct" way to be formatted (things like spacing, ordering of the keys, how to remove signatures, etc.)

The signature is particularly troublesome, since it is included inside the manifest. Most crypto signing libraries transmit the signature outside of the data that is being signed instead of inside it (think of PGP). I think that design would really make digest validation much easier. Another option is to consider a signed digest and the same digest with different or no signatures not to be the "same" so they get different digests.

mattmoor commented:
@rbarlow Very big +1, this drove us nuts because there is also [this](https://github.com/docker/distribution/blob/master/docs/spec/json.md), which would make things very straightforward.

ncdc commented on docs/spec/manifest-v2-2.md:
```diff
@@ -0,0 +1,213 @@
+- **`schemaVersion`** *int*
+	
+  This field specifies the image manifest schema version as an integer. A
+  manifest list uses version `3` to distinguish it from an image manifest.
```
This seems a bit strange to me. Doesn't this imply that if you want to rev the regular manifest version, the next available value would be 4?

mattmoor replied:
I brought this up in the DWG meeting last week as well.

I think that if these objects are going to have perpetually distinct schema versions, we should split them into distinct API objects (with their own schema number to boot).

I've been trying to articulate this for a couple meetings now, probably badly.  I will see if I can put together a proposal that better articulates it.  I think where we left things last week was to wait for Steve to weigh in this week.

ncdc commented on docs/spec/manifest-v2-2.md:
```diff
@@ -0,0 +1,213 @@
+
+    The manifests field contains a list of manifests for specific platforms.
+
+    Fields of a onject in the manifests list are:
```
s/onject/object/

stevvooe commented:
@rbarlow The content used to calculate the digest of a manifest is set by the original generator of the raw bytes. This avoids the need for specifications like Canonical JSON that purport to bring stable generation when they really don't. There are two many cross-language variations JSON generation to always get this right. This is handled simply by saving the raw bytes of the serialized manifest. This is also flexible in that it doesn't prevent implementations from generating deterministic manifests but doesn't require them where they are not needed.

rbarlow commented:
@stevvooe The problem is that signatures get inserted into the manifest, changing it. If a client needs to calculate the digest of a retrieved manifest that has signatures, that client must carefully remove the signatures and try to guess how the manifest was formatted when its digest was calculated. Since there is no canonical format for the manifest, there isn't a guarantee that doing this is possible.

If the signatures were instead stored separately of the manifest, there would be no need for any client to perform any manipulation at all on the manifest to calculate the digest. This is how PGP emails work, for example. The data being signed and the signature are separate, rather than having the signature inside the data that is being signed. This makes signature verification straightforward. The Docker manifest's inclusion of the signatures within the manifest itself make it difficult to calculate the digest of the manifest and also make signature verification difficult.

I am proposing that the digest and the signatures be separated from the manifest so that a client can calculate a plain old checksum on exactly the digest bits that were retrieved without having to do any manipulation at all. For example, something like this would be viable:

```
{'key1': 'value1', 'key2': 'value2', etc...}
------- BEGIN DOCKER SIGNATURES -------

------- END DOCKER SIGNATURES -------
```

In the example above, the client would strip the signature block and run the digest algorithm on the exact bits that are left over (the JSON). This way the client doesn't have to manipulate the JSON at all. If the original manifest used spaces and newlines or did it all in one line, the client calculates the checksum the same way - on whatever bits were received. This makes digest validation so much simpler, and it also aids in signature verification which suffers from the same problem.

Does that make sense?

aaronlehmann commented:
@rbarlow: This revision of the manifest won't include in-line signatures. Signatures will be handled outside the manifest by signing the manifest's digest, for example with Docker Content Trust.

This will be much less confusing, and I think it resolves the concern about canonical vs. non-canonical JSON.

rbarlow commented:
@aaronlehmann That is excellent news! Any hints on what sort of time frame we can expect this change to be included in Docker? I and some others are working on integrating Pulp with Docker's v2 API, and this digest calculation is currently an issue for us. Knowing a scale of time might help inform our decision about what we should do in the meantime.

aaronlehmann commented:
@rbarlow: Support for this new manifest format is planned for Docker 1.10, which will ship in a few months. I'm not sure what your compatibility plans are, but one thing I'd keep in mind is that you would need to handle the old manifest format if you want to interoperate with engines older than Docker 1.10, or registries that haven't been updated to support the new format.

I'm going to add some details on backward compatibility when I update this proposal.

aaronlehmann commented:
I've updated the proposal with the following changes:

- `s/mediaType/mediatype/`
- `schemaVersion` 2 used throughout, instead of a combination of 2 and 3
- The `label` field has been replaced by a more specific way of specifying a platform. The `platform` key maps to an object containing fields for the architecture, OS, CPU variant, and required CPU features.
- Added a backward compatibility section, explaining how the registry will support versions of the engine which don't support the new manifest format.

PTAL

RichardScothern commented:
LGTM :+1:

NathanMcCauley commented on docs/spec/manifest-v2-2.md:
```diff
@@ -0,0 +1,255 @@
+        length of the retrieved content does not match the specified length,
+        the content should not be trusted.
+    
+    - **`digest`** *string*
```
What is this digest computed over? Is it the json serialized config?

aaronlehmann replied:
Yes.

NathanMcCauley replied:
Could we say more about how it is actually constructed?

aaronlehmann replied:
IMHO the construction of this config isn't in the scope of this manifest spec.

It is pulled from the registry using the digest.

endophage replied:
It would be desirable to be able to ensure any data on disk had not been tampered with which would make it highly desirable to have a canonical format from which the digest is generated.

Is there a penalty in proscribing a specific formatting?

aaronlehmann replied:
@endophage: The image config encompasses a lot of things, such as environment variables and engine-specific configuration parameters. One of the things it includes is a list of hashes of the underlying layers. You could use these to verify the data on disk. But the config itself needs to be flexible.

endophage replied:
Having a canonical serialization format does not affect flexibility in any way. If you're saying the actual fields and values will change frequently, those changes would still necessitate serialization and hashing to get the updated digest. Requiring a strict serialization has no impact on flexibility but significant impact on reliability of digest verification.

NathanMcCauley commented on docs/spec/manifest-v2-2.md:
```diff
@@ -0,0 +1,255 @@
+        length of the retrieved content does not match the specified length,
+        the content should not be trusted.
+    
+    - **`digest`** *string*
```
Is this the digest of the gzip'd tar?

dmcgowan replied:
This would be the digest of the downloaded artifact which right now would only be gzip'd tar `application/vnd.docker.image.rootfs.diff.tar.gzip`.

endophage replied:
If deduplication of layers is a feature we want for the server, I have concerns about using digests on gzipped objects tying us down to a specific Go version (as Go carries its own gzip implementation). There is the potential that whenever a change occurs in Go's gzip implementation, all dedupe will go out the window.

This is actually worse than I've made out in the past. Being tied to a version of gzip we would have to package is one thing, being tied to a specific version of Go seems entirely unacceptable.

aaronlehmann replied:
@endophage: The v2 registry has always behaved this way. We aren't changing anything with respect to deduplication. Artifacts which compressed differently are considered distinct. It doesn't tie us to a specific version of gzip or Go at all.

endophage replied:
One of the desired features for the registry was content deduplication. If what you're saying is we don't achieve that today so who cares, let's validate that that is indeed something we can drop as a feature. Let's just make sure we get that clarity because nobody in a position to make that decision has communicated to me that we don't care about deduplication any more. Just because the implementation does achieve it now does mean we are free to move further away from achieving it, which we would be doing by locking it into what will hopefully be the new manifest format for the forseeable future.

If however deduplication remains a desired feature, and we also decide to require digests on the compressed data, it is necessary for all clients to compress data the same way, which subsequently necessitates freezing a version of gzip, and in the case of Go having its own internal gzip implementation, possibly means freezing on a Go version.

NathanMcCauley commented:
A couple of questions around workflow and integration with DCT.

1. How will the fat manifests be created? Workflow, etc.
2. If they're created on different systems, what is the expected interaction with Docker Content Trust?
3. Will DCT sign individual manifests and then sign the fat manifest from signed sub-manifests? If so, will the submanifests be able to be tagged or will we need to have a notion of signed, untagged content in DCT?

NathanMcCauley commented on docs/spec/manifest-v2-2.md:
```diff
@@ -0,0 +1,255 @@
+    
+        The MIME type of the referenced object. This will generally be
+        `application/vnd.docker.image.manifest.v2+json`, but it could also
+        be `application/vnd.docker.image.manifest.v1+json` if the manifest
```
Why allow v1 manifests at all?

aaronlehmann replied:
The "manifest lists" that point at images for each platform will be assembled with a separate tool. It may be that some of the platform-specific images it references were built with a pre-1.10 Docker engine, and therefore have an older manifest format.

Note that these are actually legacy v2 manifests. The v1 in the media type is not referring to the v1 registry protocol.

endophage replied:
I concur with Nathan. We can produce much cleaner code if we don't have to package legacy support. Would rather write microservices server side to support old clients/schemas by translating and storing in all required versions.

NathanMcCauley commented on docs/spec/manifest-v2-2.md:
```diff
@@ -0,0 +1,255 @@
+
+- **`layers`** *array*
+
+    The layer list is ordered starting from the base image (opposite order of schema1).
```
Why was the order reversed?

dmcgowan replied:
I don't remember a *good* reason why it was reversed.  With the reverse ordering the first layer seen would have the image id that used for the image and is often referred to as the "top" layer.  Layer *perception* really has no use in the manifest though so we decided base layer first makes more sense since it represents download and apply order.

aaronlehmann replied:
That's right. The current manifest format lists the top layer first, so processing over the layer list generally has to happen in reverse order. The code will be simpler if it can iterate over the layer list in the same order it appears in the manifest, so I'm proposing switching to that ordering.

stevvooe replied:
The only argument in favor or reversal is that it is easy to building when traversing from top layer down. Making assembly easy doesn't seem like a very good trade off when all the other use cases, such as pulling, have to reverse their ordering.

endophage replied:
@stevvooe wasn't that argument only relevant when images were stored as a linked list anyway?

stevvooe replied:
@endophage The daemon _does_ stores images under this model. Successive layers are accessed through traversal. The argument for top-layer first is still tenuous.

endophage replied:
Ah, I though we were proposing to change that. i.e. inspect ordered list of layers in manifest rather than traversing linked list.

estesp commented on docs/spec/manifest-v2-2.md:
```diff
@@ -0,0 +1,255 @@
+            The architecture field specifies the CPU architecture, for example
+            `x86-64` or `ppc64`.
+
+        - **`os`** *string*
```
For simplicity and ease of definition across components that would parse/utilize these fields, seems like maybe we should standardize on Golang's list here: https://golang.org/doc/install/source#environment

Note that this means variant is probably only useful for sub-architectures--which usually relates to embedded platforms like ARM, where `$GOARCH` is vague related to sub-variants.  But, it would not apply to little-endian ppc64 which has been floated up (in Linux, and Go) as a major architecture known as `ppc64le`.

For this document, if that's agreeable, all the `x86-64` references would change to `amd64`, and the note about `ppc64le` in `variant` would need to change to something about ARM, like @DieterReuter's comment on the same topic: https://github.com/docker/distribution/pull/62#issuecomment-128064649

aaronlehmann replied:
@stevvooe: What are your thoughts on this? I know we discussed this and you preferred not to tie the format to Go's architecture names.

stevvooe replied:
@estesp I agree. We should not depart from `GOARCH`/`GOOS`. They have the easiest cross-compilation system I've ever seen and would prefer to not adopt the "debian triples", if possible.

Would a few more examples help here?

I would think "variant" would the manufacturer variant field. Many of these are enumerated here: https://github.com/opencontainers/specs/issues/69#issue-94573445.

stevvooe commented on docs/spec/manifest-v2-2.md:
```diff
@@ -0,0 +1,255 @@
+            example `ppc64le` to specify a little-endian version of a PowerPC
+            CPU.
+
+        - **`features`** *array*
```
Aren't these typically called "flags"?

aaronlehmann commented:
> How will the fat manifests be created? Workflow, etc.
> If they're created on different systems (likely due to toolchain dependencies), what is the expected interaction with Docker Content Trust?

They will be created and pushed by a separate tool (possibly a Docker subcommand), after the platform-specific images are pushed. When they are pushed, they can be signed with Docker Content Trust. This might involve importing the keys, if the fat manifest is pushed from a different machine.

> Will DCT sign individual manifests and then sign the fat manifest from signed sub-manifests? If so, will the submanifests be able to be tagged or will we need to have a notion of signed, untagged content in DCT?

The tag will point at the fat manifest. The simplest use case would be to sign only the fat manifest. Since it includes digest of the image manifests, the trust would extend to those image manifests. I think it would be an interesting and useful feature to also sign the image manifest separately, but these won't have tags associated with them, so I'm not sure whether it works with the current DCT model.

stevvooe commented on docs/spec/manifest-v2-2.md:
```diff
@@ -0,0 +1,255 @@
+- `application/vnd.docker.distribution.manifest.v1+json`: schema1 (existing manifest format)
+- `application/vnd.docker.distribution.manifest.v2+json`: New image manifest format (schemaVersion = 2)
+- `application/vnd.docker.distribution.manifest.list.v2+json`: Manifest list, aka "fat manifest"
+- `application/vnd.docker.image.rootfs.diff.tar.gzip`: "Layer", as a gzipped tar
```
This should be "+gzip".

aaronlehmann replied:
The RFC says that unregistered suffixes should not be used: https://tools.ietf.org/html/rfc6838#section-4.2.8

gzip does not seem to be one of them: http://www.iana.org/assignments/media-type-structured-suffix/media-type-structured-suffix.xhtml

stevvooe replied:
Unfortunate. Do we care about the specification in this case? We've already made up "+prettyjws" and "+jws".

dmcgowan replied:
This specification does not make use of `+prettyjws` or `+jws`. The conclusion we reached after discussing was media type should be descriptive enough and `+json` is only a hint at the serialization method of the media type.

mattmoor commented on docs/spec/manifest-v2-2.md:
```diff
@@ -0,0 +1,255 @@
+
+When pushing images, clients which support the new manifest format should first
+construct a manifest in the new format. If uploading this manifest fails,
+presumably because the registry only supports the old format, the client may
```
Instead of "presumably because ..." perhaps we should instead key off of a specific rejection code, in this case: `MANIFEST_INVALID` seems appropriate.

mattmoor replied:
Can we key this behavior off of some additional metadata returned from `/v2/` instead to avoid the extra roundtrip?

Perhaps we can make `/v2/` start to return `Accept` headers?

mattmoor commented on docs/spec/manifest-v2-2.md:
```diff
@@ -0,0 +1,255 @@
+When pulling images, clients indicate support for this new version of the
+manifest format by sending the
+`application/vnd.docker.distribution.manifest.v2+json` and
+`application/vnd.docker.distribution.manifest.list.v2+json` media types in an
```
Shouldn't this include a mimetype for the legacy format?

Can we illustrate it via a sample HTTP request maybe?

stevvooe replied:
@mattmoor That seems outside the scope of this specification. Should we supply the modifications to the v2 api specification concurrently?

mattmoor replied:
Sure, I'm fine with that.  I just think something a bit more tangible (as exists in api.md) would be helpful.

mattmoor commented on docs/spec/manifest-v2-2.md:
```diff
@@ -0,0 +1,255 @@
+manifest format by sending the
+`application/vnd.docker.distribution.manifest.v2+json` and
+`application/vnd.docker.distribution.manifest.list.v2+json` media types in an
+`Accept` header when making a request to the `manifests` endpoint. Updated
```
Can we reflect this with an extension to docs/specs/api.md?

mattmoor commented on docs/spec/manifest-v2-2.md:
```diff
@@ -0,0 +1,255 @@
+If the manifest being requested uses the new format, and the appropriate media
+type is not present in an `Accept` header, the registry will assume that the
+client cannot handle the manifest as-is, and rewrite it on the fly into the old
+format. If the object that would otherwise be returned is a manifest list, the
```
I think an example for each case (stepping through the transformation) would clarify the expected process.

mattmoor commented on docs/spec/manifest-v2-2.md:
```diff
@@ -0,0 +1,255 @@
+registry will look up the appropriate manifest for the x86-64 platform and
+linux OS, rewrite that manifest into the old format if necessary, and return
+the result to the client. If no suitable manifest is found in the manifest
+list, the registry will return a 404 error.
```
It is worth noting that manifest/schema1/sign.go specifically states:
```go
// Sign signs the manifest with the provided private key, returning a
// SignedManifest. This typically won't be used within the registry, except
// for testing.
```

However, signing will be very necessary on the serving path in this model.

mattmoor commented on docs/spec/manifest-v2-2.md:
```diff
@@ -0,0 +1,255 @@
+new format only provides one image configuration. To work around this, the
+registry will create synthetic image configurations for all layers except the
+top layer. These image configurations will not result in runnable images on
+their own, but only serve to fill in the parent chain in a compatible way.
```
Again, an example would be useful in helping illustrate this.  Perhaps include a skeleton template we should use?

mattmoor commented on docs/spec/manifest-v2-2.md:
```diff
@@ -0,0 +1,255 @@
+The IDs in these synthetic configurations will be derived from hashes of their
+respective blobs. The registry will create these configurations and their IDs
+using the same scheme as Docker 1.10 when it creates a legacy manifest to push
+to a registry which doesn't support the new format.
```
Can you link to a specification for this for those of us that haven't seen the 1.10 plans?

stevvooe commented:
@mattmoor I agree that we need some clarification added to the V2 api specification, but I don't want to hold up this specification while waiting on the other specification changes.

@aaronlehmann Does taht sound okay?

aaronlehmann commented:
On Wed, Oct 21, 2015 at 12:44:30PM -0700, Stephen Day wrote:
> @mattmoor I agree that we need some clarification added to the V2 api
> specification, but I don't want to hold up this specification while
> waiting on the other specification changes.
> 
> @aaronlehmann Does taht sound okay?

That sounds good to me.

endophage commented:
Overall this looks like a significant improvement to the current status quo. As far as DCT, it would appear continuing to sign tags would be sufficient to maintain the current level of security. We would still have the same merkel tree based guarantees and DCT wouldn't have to know anything about whether it's a "fat" manifest or otherwise, the client would work that out based on content type headers.

endophage commented:
Also, my 2 cents on JSON serialization. It doesn't have to be the Canonical JSON spec, but please, please, prescribe a format. At an absolute minimum, specify an encoding (yes the JSON spec already does this, I've still seen people manage to put Windows-1252 into JSON, just reiterate the JSON spec), specify removal of all whitespace (we're going to be operating on Windows and *nix, I see "\n" vs. "\r\n" problems in our future if we don't do this), and specify fields are ordered lexically (the way Go already does it).

I said it before, there's no penalty to doing this and consistency provides significant benefits in our ability to reliably re-verify JSON data regardless of how many cycles of parsing - serialization it goes through on any number of tools and OSes.

stevvooe commented:
@endophage If only we had a [json specification](https://github.com/docker/distribution/blob/master/docs/spec/json.md)...

endophage commented:
@stevvooe If only this spec referenced that as the json format it required...

stevvooe commented:
@endophage :smile_cat:

tiborvass commented on docs/spec/manifest-v2-2.md:
```diff
@@ -0,0 +1,255 @@
+and is specified in the [schema 1 manifest definition](./manifest-v2-1.md)
+
+This second schema version has two primary goals. The first is to allow
+multi-architecture images, through a "fat manifest" which references image
```
@aaronlehmann We're discussing this PR with @coolljt0725 and we were thinking about `docker build`. How would `docker build` create a multi-arch image? If not on build, then on push (that's a detail). Now if we say "it only creates arch-specific manifests", then we should have a way of creating a arch-independent manifest (sort of a meta-manifest) that references the arch-specific manifests. If that makes  sense, that could be an argument against fat manifests. WDYT?

tiborvass replied:
Or at least, we need a way of providing the option of having manifest-referencing manifests, in addition to fat manifests.

estesp replied:
`docker build` will work as it does today creating a single architecture image.  `docker push` is what creates manifests today, and the discussion will have to happen around whether we can "tuck" the features of creating a multi-arch manifest (and reference/push layers for more than one architecture image) in the flow of `docker push` or will `docker push` create a traditional single-architecture manifest, and that digest hash will be used in a command like `docker manifest [create]`, along with other digest hashes to construct the 'fat manifest' after all single-architecture manifests/layers are pushed?

We had a discussion about this briefly with @RichardScothern and @stevvooe here at DockerCon, and I hope to, with @harche, create some user flows and add them as a comment in #200 (or wherever makes sense) as it is broader than just the manifest schema format itself--which is this PR's focus.  Once we have those, that should give us a framework of actions/proposed commands/user flow to have a targeted discussion on what makes the most sense.

tiborvass replied:
My misunderstanding came from the name `fat manifest`, I thought all the arch-specific sets of layers would be in one manifest. Instead, this described a manifest with a list of manifests. I would suggest to change the name. A manifest can either have a set of layers, or a set of manifests. However, that would allow having manifests of manifests of manifests.

stevvooe replied:
I don't really think `docker build` makes sense for multi-arch.

> A manifest can either have a set of layers, or a set of manifests. However, that would allow having manifests of manifests of manifests.

The original design from this phase said manifests enumerated resources. Those resources could be layers or other manifests. I am not sure why we departed from this.

jfrazelle commented:
This would be really helpful for upgrading our download-frozen-image script from v1, ping @tianon

davedoesdev commented on docs/spec/manifest-v2-2.md:
```diff
@@ -0,0 +1,255 @@
+# Image Manifest
+
+The image manifest provides a configuration and a set of layers for a container
+image. It's the direct replacement for the schema-1 manifest.
```
What if I don't want to store container images? I just want to add some metadata about some blobs.
Is that supported?

davedoesdev commented on docs/spec/manifest-v2-2.md:
```diff
@@ -0,0 +1,255 @@
+    
+    - **`size`** *int*
+    
+        The size in bytes of the object. This field exists so that a client
```
Is this optional? Otherwise in order to add a manifest referencing a number of 'layers' (or blobs), after they've been pushed at some earlier time, I'll have to go get their sizes.
I'm not sure what security benefits checking the lengths actually achieves.

aaronlehmann replied:
If you are storing a record of having pushed the layer previously, can you also store the size in that record? Otherwise, I'd think you would need to `Stat` the blob to verify that it exists on the remote size, and this returns a descriptor that includes the size.

davedoesdev replied:
I can do either, yes. But I don't always _need_ the size.

davedoesdev replied:
Btw, what do you by `Stat`? If I'm using the REST API then I should be able to do a `HEAD` request to get the blob's size, right?

aaronlehmann replied:
Yes, that's right. The typical flow when pushing a layer is to do a `HEAD` request first to see whether it already exists.

davedoesdev commented:
Not sure if this is the right place for this question but why does the distribution server even care about the manifest format?
Why not treat the manifest as an opaque blob and just give that back to clients? Perhaps you'd also store a mime type supplied on upload but that's all.
Kind of like any other blob upload but instead of being content addressable it'd have a name (via `/manifests/xxx`).

mattmoor commented:
@davedoesdev With the goal of strictly serving v2 (roundtripping metadata as-is), I think I'd agree that this would be a fantastic separation of concerns.  Big +1.


Here are a few interesting complications in practice:
  1. Serving v1 from v2 data (some of us still do)
  1. Pull by digest the current manifest format's digest isn't sha256sum(what's passed as manifest)

Say we were to fix the second point with this spec, well then we'd still need to serve the legacy v2 format (and/or v1 format) to older clients for a period.


Honestly, the only two things I care about getting out of this manifest do-over are:
  1. sha256sum(input) = the content-addressable name.
  1. A clear compatibility story (possibly requiring a server-side slow-path for serving v1 and older v2 clients).

My $0.02...

davedoesdev commented:
@mattmoor makes sense and thanks

aaronlehmann commented:
Good summary from @mattmoor. If this design was happening from scratch, we would make the registry treat the manifest as an opaque blob. But we have to live with the existing manifest format, where the content address is not generated from a simple SHA256 over the content. And we also want to provide backward compatibility for old engines by rewriting manifests in the new format when necessary.

mattmoor commented:
@aaronlehmann Something that was bugging me last night (wrt compat), which I think should be called out in the compat section of this proposal:

What happens when I push with a new client, which would spit out a digest: `sha256:deadbeaf`.

Then I tell my cluster (possibly running an older version) to pull `image@sha256:deadbeef`.

There are two possible outcomes:
  * The cluster speaks the new format, or
  * What it pulls (back-compat) won't match the digest.

I think the only remediation in this case (old client digest-pulling new manifest) is for digest-based pulls returning a 400 and diagnostic that is properly displayed by 1.6+ clients.  thoughts?

stevvooe commented:
> Not sure if this is the right place for this question but why does the distribution server even care about the manifest format?

@davedoesdev In general, @mattmoor's commentary is all right on. I'll try to expand on that with some history. (A read over the ROADMAP.md may help, as well).

Ideally, the registry would only speak blobs. This model is ultimately superior, as it allows docker to iterate without considering the remote storage format. Unfortunately, several features were placed in the charge of the registry, such as tagging (ie naming) and search. These feature need to have a semantic understanding of the data to be useful. This has lead users to feel these features are important for a _complete_ registry. 

With the design of V2, we have had a larger focus on breaking up the feature set into components. While we have gone very far in this direction, we must be careful not to stretch the bounds of backwards compatibility too far. The main "stretch" here is that the V1 protocol relied heavily on trusting naming and identification to the remote registry, whereas, in V2, this all lies on the daemon integrated with several services.

Ultimately, there needs to some common format between the client code, part of distribution, and the registry to instruct the distribution component on what to fetch. In git, references, commits and trees play this role. For docker, the manifest plays this role. For the purpose of de-coupling, the more generic the better. In fact, if you go back and look at previous versions of this proposal, they were actually much more generic. This was changed, as many reviewers didn't understand the need for a generic format (@aaronlehmann can comment further on this, I may be oversimplifying). Either way, this solution gets us to the point where we can have decent content-addressability and strong integration points to give us protocol flexibility.

Digressing, most of what you see here is to support the gymnastics required to move to a fully content-addressable system. While the registry will only get dumber, it still needs to have a level functionality to fill in the behavior of past registries. It may look weird but I am not sure if any transition from an open system to a content-addressable system won't.

I hope this helps to clarify some points.

aaronlehmann commented:
@mattmoor: You're correct that this isn't a use case we can support directly. Existing engines check the manifest digests when they pull by digest, so they will fail to pull schema2 manifests by digest. I'll add some discussion about this to the document.

davedoesdev commented:
Great replies, thanks again - they clear it up nicely.

endophage commented:
I had to spend the weekend thinking about how to best respond to this verification discussion. I strongly disagree with the position here. There are a lot of reasons why:

1. Currently, a client can push a bad manifest and the registry will happily accept it. When any client then attempts to pull it, it will fail. This makes it appear like the registry has corrupted the manifest. This also means N pullers may fail, as opposed to _one_ pusher.
2. The registry is an entry point for a cascade of failures. Various backend systems like autobuilds and nautilus fail noisily (for the on-calls) and in non-obvious ways (the user will probably not know what's wrong) when the registry accepts bad manifests.
3. If the distribution team doesn't care about the format, they shouldn't be specifying it. If however you want to dictate a format, it's up to you to enforce it.
4. The above points make the registry (currently) a bad citizen. The distribution team is the only team not getting paged about bad data the registry has allowed into the system.
5. I'm sure somebody will ask: "Then should we be verifying tars? That's unreasonable!" No, I'm not advocating tars for one simple reason: tars have a well understood input and output formats with tooling and libraries for near every OS and language to help users generate sufficiently valid tar files. Manifests do not have such tooling, and people do build their own generators.

aaronlehmann commented:
@endophage: I feel like I'm missing some context. Which comments are you referring to, and what verification steps do you have in mind for the new manifest format being discussed here?

endophage commented:
@aaronlehmann the 8 comments immediately preceding mine are a discussion that appears to have the conclusion of "the registry should just accept opaque blobs."

davedoesdev commented:
@endophage I'm happy for the registry to validate mime types it understands.
I was under the impression that the registry was moving towards being a more general system for content distribution. In which case it'd be great if it could also support 'opaque' manifest types.

endophage commented:
@davedoesdev re "validate mime types it understands": I'm happy with that as a compromise as long as those mime types include all manifest JSON objects.

mattmoor commented:
@endophage opaque doesn't mean unverifiable digests.  I don't need to understand a mime type to compute it's sha256.  However, today that is exactly what's required to compute the sha256 of manifests.  I'd much rather have the content-addressing be derived in a content-opaque fashion, regardless of what deeper understanding the registry has of the content.

endophage commented:
@mattmoor I have no issue with that. Verification is different from addressing. I just want to be clear though that the registry needs to verify manifests during upload and provide synchronous responses when they're invalid.

aaronlehmann commented:
I don't think the previous conversation mentioned verification, which is why I was confused.

Despite the philosophical discussion about how it might be better if blobs were opaque, the registry parses the manifest JSON, and would return an error to the client if it's not valid.

For the new manifest format, the only other sensible verification step I can think of would be to make sure that the other digests referenced by the manifest exist in storage. But I'm not 100% sure we should do this - there may be use cases where someone would want to upload a manifest first, and then its dependencies, or situations where eventual consistency leads to the dependencies not being immediately visible. Any thoughts?

endophage commented:
@aaronlehmann I wish that was true but beyond side effect verification (errors attempting to deserialize JSON and digests) there is very little done at the moment. It would be beneficial to also validate that the schema version is an acceptable/known version, if all required fields for that version are present, if integer fields, like size, actually contain integers. The new format appears relatively simple so those are the only things that spring to mind at the moment.

stevvooe commented:
@endophage This approach to content verification and validation on the registry side has resulted in a massively broken security situation and an inflexible distribution story. The dumber we make the registry, the more protected the daemon must become.

harche commented:
@aaronlehmann May be I am missing something, but how can someone just upload this new manifest format (Fat manifest) and then later the dependencies (I assume you meant the other manifests it references)? Won't that change the Fat manifest's digest? 

IMO, this should behave like a pure content addressable storage system (something like IPFS). Users can create these Fat manifests based on existing manifest digests. If someone needs to add another reference manifest then they just have to create new Fat manifest with all the older references plus the additional one. 

If one of the manifest reference in Fat manifest gets deleted, the error should occur only when user is trying to access the reference manifest through Fat manifest. This will let user know that this Fat manifest should not be used anymore because the reference manifest they are interested in is no longer available. This way no logic has to be present in registry at all and it can behave like a dumb blobstore.

aaronlehmann commented:
I was only thinking that the uploads might be ordered in a way where the dependencies can't be checked immediately. This probably won't be a problem in practice. BTW, the code that's in progress for the Docker registry validates dependencies referenced by the manifest, with an exception for the proxy mode.

harche commented:
@aaronlehmann Interesting and what does it do if validation of the dependencies fail? Does it validate when the fat manifest is getting created or when it is being pulled by daemon?

harche commented:
@aaronlehmann @stevvooe @estesp Should I create a separate issue to discuss how the CLI going to look like for fat manifests?

stevvooe commented:
@harche Yes. This would probably make the most sense. We can have the proposal stage in the distribution project for now and dispatch to docker/docker depending on where the actual feature will land.

aaronlehmann commented:
I've updated this proposal to make a few minor changes:

- Include the content type in-band in both image manifests and manifest lists. This makes it easy to distinguish between image manifests and manifests lists in case there's no content type available out-of-band.
- Changed mediatype to mediaType, because this is the way the distribution codebase already serializes Descriptor objects, and I don't want to deviate from that. If people prefer to change this serialization (and it's safe to do so), I can change it back.

I feel like this is ready to be merged. PTAL.

I have working code that implements the formats in this proposal, which I'll open as a PR against `docker/distribution` once #1268 is merged.

estesp commented:
Thanks @aaronlehmann ; looks good and glad to see the other PRs showing up with implementation.  I'm "non-maintainer LGTM" on this one! :)

RichardScothern commented:
LGTM

On Thu, Dec 17, 2015 at 8:17 PM, Phil Estes <notifications@github.com>
wrote:

> Thanks @aaronlehmann <https://github.com/aaronlehmann> ; looks good and
> glad to see the other PRs showing up with implementation. I'm
> "non-maintainer LGTM" on this one! :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/distribution/pull/1068#issuecomment-165669758>.
>

stevvooe commented:
> Include the content type in-band in both image manifests and manifest lists. This makes it easy to distinguish between image manifests and manifests lists in case there's no content type available out-of-band.

Content-type should always be in-band, from the trusted source. Content-type returned by registries (out-of-band) should not be trusted. We might want to say this explicitly, as this is important for security.

> Changed mediatype to mediaType, because this is the way the distribution codebase already serializes Descriptor objects, and I don't want to deviate from that. If people prefer to change this serialization (and it's safe to do so), I can change it back.

Please don't. Leave as "mediatype". The fact that descriptor has camelCase is unintentional. In general, keys should be case-insensitive.

aaronlehmann commented:
> Please don't. Leave as "mediatype". The fact that descriptor has camelCase is unintentional. In general, keys should be case-insensitive.

Would you like me to change it to `mediatype` in descriptor as well? My implementation of the schema2 manifest uses `distribution.Descriptor` directly, and I would like to avoid forking it.

stevvooe commented:
> Would you like me to change it to mediatype in descriptor as well? My implementation of the schema2 manifest uses distribution.Descriptor directly, and I would like to avoid forking it

@aaronlehmann I don't think we use that anywhere, so go ahead.

aaronlehmann commented:
Reverted the `mediatype` change; will change the `Descriptor` serialization to match in #1281.

aaronlehmann commented:
@stevvooe: Changing the `Descriptor` serialization would change affect the notification serialization format. Are we confident enough that implementation will be case-insensitive to change the casing there?

stevvooe commented:
@aaronlehmann I guess we can leave it alone and change to `mediaType` in this specification. We have no guarantee that consumers wrote appropriately agile code to handle case-insensitivity. Sorry for the confusion.

aaronlehmann commented:
Changed back. No worries.

dmp42 commented:
LGTM

stevvooe commented:
LGTM
